src/irmin/object_graph.ml

```diff
@@ -156,32 +157,37 @@ struct
     let rec visit () =
       match Queue.pop todo with
       | exception Queue.Empty -> return_unit
-      | key, level ->
+      | key, level -> (
           if level >= depth then visit ()
           else if has_mark key then visit ()
-          else (
-            mark key level;
-            Log.debug (fun f -> f "VISIT %a %d" Type.(pp X.t) key level);
-            f_nodes key >>= fun () ->
-            pred key >>= fun keys ->
-            List.iter (fun k -> Queue.push (k, level + 1) todo) keys;
-            f_edges key keys >>= fun () -> visit () )
+          else
+            skip key >>= function
```

Comment:
it's probably a good idea to mark the node here so `skip` is called only once for every node.